### PR TITLE
Pull out a dart_vm_binary_action

### PR DESCRIPTION
**Pull request by natebosch**
- Rename dart_vm_binary_impl as dart_vm_binary_action and make the
  public attributes arguments so it can be reused form other rules.
- Add dart_vm_binary_defaults to add the attributes with default values
  to a dict.
- Move the rule _impl to the file defining the rule as a thin wrapper
  around the action to pull out the attributes.
- Add a struct 'dart_vm' to collect the identifiers that will be useful
  to load together

Cleanup:
- Add licence preamble to dart_vm_binary.bzl

### CODE REVIEW

**Comment by nshahan** [comment]
(nit) Fix typo in the commit message: dart_vm_binary_defautls -> dart_vm_binary_defaults

**Comment by natebosch** [comment]
Good catch - will fix when I merge

File: dart/build_rules/vm.bzl [hunk 1]
```diff
@@ -97,3 +111,8 @@ dart_vm_test = rule(
     test = True,
     implementation = dart_vm_test_impl,
 )
+
+dart_vm = struct(
```
Comment (jakemac53) [inline dart/build_rules/vm.bzl]: Do we need to export these? Should we add "binary" and/or "snapshot"?

Comment (natebosch) [inline dart/build_rules/vm.bzl]: `dart_vm` matches what we have internally. The 'binary' part comes in the keys inside the struct.

When we have use case for using snapshot_action etc in rules that are externalized we'll either do `dart_vm.snapshot_action` or add another layer of nesting like `dart_vm.binary.action`

**Pull request merged**